### PR TITLE
Fix DDP Reduction in PyTorch 1.7 when the model has unused parameters

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation and Context

PyTorch 1.7 slightly changed the way Reducer should be used when the model has unused parameters, leading to DD-PPO breaking when the model does, why this is certainly an edge case, it is a problem.  

This PR solve this and attempts to solve future issues by using a wrapper to interface with DDP's forward call instead of using the undocumented Reducer APIs.  While I am not sure if this won't cause it's own issues since it is kinda hacky, this seems like it has the potential to be better.

## How Has This Been Tested

Tested on a model with unused parameters.  I will write a test for this before merging this, but don't have time today.

## Types of changes

- Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 1]
```diff
@@ -79,24 +91,25 @@ class Guard:
...
 
-    def before_backward(self, loss: Tensor) -> None:
-        super().before_backward(loss)  # type: ignore
+        self._evaluate_actions = Guard(_EvalActionsWrapper(self.actor_critic), self.device)  # type: ignore
 
-        if self.find_unused_params:
-            self.reducer.prepare_for_backward([loss])  # type: ignore
-        else:
-            self.reducer.prepare_for_backward([])  # type: ignore
+    def evaluate_actions(
```
Comment (mathfac) [inline habitat_baselines/rl/ddppo/algo/ddppo.py:107]: Doc string will be useful.

File: habitat_baselines/rl/ppo/ppo.py [hunk 2]
```diff
@@ -152,6 +152,13 @@ def update(self, rollouts: RolloutStorage) -> Tuple[float, float, float]:
 
         return value_loss_epoch, action_loss_epoch, dist_entropy_epoch
 
+    def evaluate_actions(
```
Comment (mathfac) [inline habitat_baselines/rl/ppo/ppo.py:155]: Doc string will be useful.

**Review by mathfac — APPROVED** [review]

File: habitat_baselines/rl/ppo/ppo.py (hunk 2 above)
Comment (erikwijmans) [reply habitat_baselines/rl/ppo/ppo.py]: Add a little doc string and made it internal since it just exists to be overridden with inheritance.

**Comment by erikwijmans** [comment]
@mathfac @Skylion007 does this test look reasonable?  It's a little ugly but I wanted it to be as close to testing the actual code path used in training as possible.

**Review by Skylion007 — APPROVED** [review]

**Comment by mathfac** [comment]
Yes, looks good.

**Pull request merged**